bootstore/src/schemes/v0/request_manager.rs

```diff
@@ -0,0 +1,401 @@
+    pub received: BTreeMap<Baseboard, Share>,
+}
+
+impl ShareAcks {
+    pub fn new(threshold: u8) -> ShareAcks {
+        ShareAcks { threshold, received: BTreeMap::new() }
+    }
+}
+
+/// A mechanism to track in flight requests
+///
+/// Manages expiry, acknowledgment, and retries
+#[derive(Debug)]
+pub enum TrackableRequest {
+    /// A request from the caller of the Fsm API to initialize a rack
+    ///
+    /// This must only be called at one peer, exactly once. That peer
+    /// will be in `InitialMember` state.
+    InitRack {
+        rack_uuid: Uuid,
+        packages: BTreeMap<Baseboard, SharePkg>,
+        acks: InitAcks,
+    },
+
+    /// A request from the caller of the Fsm API to load a rack secret
+    ///
+    /// Only peers in `InitialMember` or `Learned` state can load rack secrets
+    LoadRackSecret { rack_uuid: Uuid, acks: ShareAcks },
+
+    /// A request received from a peer to learn a new share
+    ///
+    /// This request results in share gathering so "extra" shares can be
+    /// decrypted via a key derived from the rack secret and one of them
+    /// distributed to the learner.
+    ///
+    /// Only peers in `InitialMember` state can respond successfully
+    LearnReceived { rack_uuid: Uuid, from: Baseboard, acks: ShareAcks },
+
+    /// A request sent from a peer in `Learning` state to another peer
+    /// to learn a key share.
+    LearnSent { to: Baseboard },
+}
+
+/// A mechanism to manage all in flight requests
+///
+/// We expect very few requests at a time - on the order of one or two requests.
+pub struct RequestManager {
+    id: Baseboard,
+    config: Config,
+    requests: BTreeMap<Uuid, TrackableRequest>,
+    expiry_to_id: BTreeMap<Instant, Uuid>,
+
+    /// Messages that need sending to other peers.
+    ///
+    /// These should be drained on each API call.
+    envelopes: Vec<Envelope>,
+}
+
+impl RequestManager {
+    /// Create a new RequestManager
+    pub fn new(id: Baseboard, config: Config) -> RequestManager {
+        RequestManager {
+            id,
+            config,
+            requests: BTreeMap::new(),
+            expiry_to_id: BTreeMap::new(),
+            envelopes: vec![],
+        }
+    }
+
+    /// Return an iterator of drained envelopes
+    pub fn drain_elements(&mut self) -> impl Iterator<Item = Envelope> + '_ {
+        self.envelopes.drain(..)
+    }
+
+    /// Track a new `Fsm::init_rack` api request and broadcast a
+    /// `RequestType::Init` to connected peers.
+    pub fn new_init_rack_req(
+        &mut self,
+        now: Instant,
+        rack_uuid: Uuid,
+        packages: BTreeMap<Baseboard, SharePkg>,
+        connected_peers: &BTreeSet<Baseboard>,
+    ) -> Uuid {
+        let expiry = now + self.config.rack_init_timeout;
+        let mut acks = InitAcks::default();
+        acks.expected = packages
+            .keys()
+            .cloned()
+            .filter_map(|id| if id != self.id { Some(id) } else { None })
+            .collect();
```

Comment:
`filter_map` seems unnecessary here:
```suggestion
        acks.expected =
            packages.keys().cloned().filter(|id| id != &self.id).collect();
```

Reply:
Duh! Thanks!